models/marts/dim_contract_data_hist.sql

```diff
@@ -0,0 +1,216 @@
+            , s.asset_code as source_asset_code
+            , t.asset_code as target_asset_code
+            , s.asset_issuer as source_asset_issuer
+            , t.asset_issuer as target_asset_issuer
+            , s.asset_type as source_asset_type
+            , t.asset_type as target_asset_type
+            , s.balance as source_balance
+            , t.balance as target_balance
+            , s.balance_holder as source_balance_holder
+            , t.balance_holder as target_balance_holder
+            , s.contract_create_ts as source_contract_create_ts
+            , t.contract_create_ts as target_contract_create_ts
+            , s.contract_delete_ts as source_contract_delete_ts
+            , t.contract_delete_ts as target_contract_delete_ts
+            , s.closed_at as source_closed_at
+            , t.closed_at as target_closed_at
+            , t.end_date as target_end_date
+            , t.is_current as target_is_current
+            , s.airflow_start_ts as source_airflow_start_ts
+            , t.airflow_start_ts as target_airflow_start_ts
+            , s.batch_id as source_batch_id
+            , t.batch_id as target_batch_id
+            , s.batch_run_date as source_batch_run_date
+            , t.batch_run_date as target_batch_run_date
+            , s.row_hash as source_row_hash
+            , t.row_hash as target_row_hash
+            , t.dw_load_ts as target_dw_load_ts
+            , t.dw_update_ts as target_dw_update_ts
+            , case
+                when t.ledger_key_hash is null then 'Insert'  -- New record
+                when s.row_hash != t.row_hash then 'Update'  -- Changed record
+                else 'NoChange'  -- No change in the current record
+            end as change_type
+        from source_data as s
+        full outer join target_data as t
+            on s.ledger_key_hash = t.ledger_key_hash
+            and s.start_date = t.start_date
+    )
+
+    -- Date chaining for SCD Type 2 with separated CTEs
+    , date_chained as (
+        select
+            cdc.*
+            , lead(cdc.start_date) over (partition by cdc.ledger_key_hash order by cdc.start_date) as next_start_date
+            -- Operation Types:
+            -- INSERT_NEW_KEY: First occurrence of a ledger_key_hash
+            -- START_NEW_VERSION: New version of existing record due to attribute changes
+            -- END_CURRENT_VERSION: Close current version due to future changes
+            -- KEEP_CURRENT: No changes needed, maintain current record
+            , case
+                when cdc.change_type = 'Insert' then 'INSERT_NEW_KEY'
+                when cdc.change_type = 'Update' then 'START_NEW_VERSION'
+                when
+                    cdc.change_type = 'NoChange'
+                    and lead(cdc.start_date) over (partition by cdc.ledger_key_hash order by cdc.start_date) is not null
+                    then 'END_CURRENT_VERSION'
+                else 'KEEP_CURRENT'
+            end as operation_type
```

Comment:
Do you need to create/track the change_type? It looks like any change is processed the same way

Reply:
- Can you please elaborate what you mean. I didn't get it.

Reply:
You name mark rows with `change_type` and `operation_type` but never use this downstream. At the end of the scd2 table you don't include `change_type` nor `operation_type` so is there a point in identifying a change with such detail if all you need is a boolean `is_changed`

Reply:
Understood. Agree that everything is either an Insert/Upsert but having the rows explicitly marked help identify things during debugging/testing.